syntax = "proto3";

package weather;

service WeatherService {
  rpc GetWeatherData(WeatherDataRequest) returns (WeatherDataResponse);
}

message Coordinates {
  double latitude = 1;
  double longitude = 2;
}

message WeatherDataRequest {
  Coordinates coordinates = 1;
  repeated string exclude = 2; // Sections to exclude: current, minutely, hourly, daily, alerts
  string units = 3; // standard, metric, imperial
  string language = 4; // Language code
}

message WeatherDataResponse {
  CurrentWeather current = 1;
  repeated MinuteForecast minutely = 2;
  repeated HourlyForecast hourly = 3;
  repeated DailyForecast daily = 4;
  repeated WeatherAlert alerts = 5;
}

message WeatherCondition {
  int32 id = 1;
  string main = 2;
  string description = 3;
  string icon = 4;
}

message CurrentWeather {
  int64 dt = 1;
  int64 sunrise = 2;
  int64 sunset = 3;
  double temp = 4;
  double feels_like = 5;
  int32 pressure = 6;
  int32 humidity = 7;
  double dew_point = 8;
  double uvi = 9;
  int32 clouds = 10;
  int32 visibility = 11;
  double wind_speed = 12;
  int32 wind_deg = 13;
  double wind_gust = 14;
  repeated WeatherCondition weather = 15;
}

message MinuteForecast {
  int64 dt = 1;
  double precipitation = 2;
}

message HourlyForecast {
  int64 dt = 1;
  double temp = 2;
  double feels_like = 3;
  int32 pressure = 4;
  int32 humidity = 5;
  double dew_point = 6;
  int32 clouds = 7;
  int32 visibility = 8;
  double wind_speed = 9;
  int32 wind_deg = 10;
  double wind_gust = 11;
  double pop = 12;
  repeated WeatherCondition weather = 13;
}

message DailyForecast {
  int64 dt = 1;
  int64 sunrise = 2;
  int64 sunset = 3;
  int64 moonrise = 4;
  int64 moonset = 5;
  double moon_phase = 6;
  Temperature temp = 7;
  FeelsLike feels_like = 8;
  int32 pressure = 9;
  int32 humidity = 10;
  double dew_point = 11;
  double wind_speed = 12;
  int32 wind_deg = 13;
  double wind_gust = 14;
  int32 clouds = 15;
  double pop = 16;
  double uvi = 17;
  repeated WeatherCondition weather = 18;
}

message Temperature {
  double day = 1;
  double min = 2;
  double max = 3;
  double night = 4;
  double eve = 5;
  double morn = 6;
}

message FeelsLike {
  double day = 1;
  double night = 2;
  double eve = 3;
  double morn = 4;
}

message WeatherAlert {
  string sender_name = 1;
  string event = 2;
  int64 start = 3;
  int64 end = 4;
  string description = 5;
  repeated string tags = 6;
}